.github/workflows/auto-merge.yml

```diff
@@ -20,23 +18,38 @@ concurrency:
 
 jobs:
   dependabot:
+    name: Dependabot PR auto-merge
     runs-on: ubuntu-latest
     if: ${{ github.actor == 'dependabot[bot]' }}
     steps:
-      - name: Dependabot metadata
-        id: metadata
-        uses: dependabot/fetch-metadata@v1.3.1
-        with:
-          github-token: "${{ secrets.GITHUB_TOKEN }}"
-      - name: Enable auto-merge for Dependabot PRs
-        if: ${{ github.event.pull_request.auto_merge == null && ( steps.metadata.outputs.update-type == 'version-update:semver-patch' || steps.metadata.outputs.update-type == 'version-update:semver-minor' ) }}
-        run: gh pr merge --auto --merge "$PR_URL"
+      - name: Checkout
+        uses: actions/checkout@v3
+
+      - name: Approve and merge the PR
+        if: |
+          contains( github.event.workflow_run.head_commit.message, 'version-update:semver-minor' ) ||
+          contains( github.event.workflow_run.head_commit.message, 'version-update:semver-patch' )
+        run: |
+          gh pr review "$PR_NUMBER" --approve --body "LGTM!"
+          gh pr merge "$PR_NUMBER" --auto --merge --delete-branch
         env:
-          PR_URL: ${{github.event.pull_request.html_url}}
+          PR_NUMBER: ${{github.event.workflow_run.pull_requests[0].number}}
           GITHUB_TOKEN: ${{secrets.GITHUB_TOKEN}}
-      - name: Approve PR
-        run: gh pr review --approve "$PR_URL"
-        if: ${{ steps.metadata.outputs.update-type == 'version-update:semver-patch' || steps.metadata.outputs.update-type == 'version-update:semver-minor' }}
+
+  gutenberg-packages:
+    name: Gutenberg packages update PR auto-merge
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+
+      - name: Approve and merge the PR
+        if: |
+          startsWith(github.event.workflow_run.head_commit.message, 'Update WordPress NPM packages based on Gutenberg release') &&
+          endsWith(github.event.workflow_run.head_commit.message, format('{0} {1}', 'github-actions', '<github-actions@github.com>')) &&
+          github.event.workflow_run.head_commit.author.name == 'github-actions' &&
+          github.event.workflow_run.head_commit.author.email == 'github-actions@github.com'
+        run: gh pr merge "$PR_NUMBER" --auto --merge --delete-branch
```

Comment:
What would happen if someone opened a PR with a malicious commit that has the name and author set to `github-actions`? Would this allow them to get anything merged automatically? It would seem safer to do the merge only in the `gutenberg-packages-update.yml` workflow after we know that it was this same workflow that was the one to have made the changes.

Reply:
Good catch! Thanks.

> It would seem safer to do the merge only in the `gutenberg-packages-update`.yml workflow

I didn't find a perfect event for this yet, like waiting to pass all the checks. `workflow_run` event is the perfect to trigger `auto-merge`.

As far as we are concerned about malicious commits we can eliminate this by adding `if: ${{ github.actor == 'PAT_USER' }}` on the job. `PAT_USER` is the username of the user who has provided the `personal access token` in

https://github.com/ampproject/amp-wp/blob/0754d926916d62b2339198b97553d50363260779/.github/workflows/gutenberg-packages-update.yml#L187

Reply:
I fully admit that I am not an expert on GHA, so pardon my ignorance. As I understand, this workflow will run for every pull request. In this case, the `actor` could be Dependabot, a contributor user, or the system cron (in the case of scheduled Gutenberg updates). I'm not clear on how `github.actor` would be the same as the PAT user unless the specific GitHub user opened the PR.

Reply:
Actually, my thinking process here is to avoid merging commits from malicious users. 

Let's say you add your `personal access token` in the above `GH_PAT` so `github.actor` will be `westonruter` because the workflow will be triggered from your `PAT`, hence I am certain that a commit with `author` as `github-action` can't be merged in the repo. Workflow will be having two checks in this condition.

```yaml
gutenberg-packages:
  name: Gutenberg packages update PR auto-merge
  runs-on: ubuntu-latest
  if: ${{ github.actor == 'trusted-user' }}
  steps:
    - name: Checkout
      uses: actions/checkout@v3

    - name: Approve and merge the PR
      if: |
        startsWith(github.event.workflow_run.head_commit.message, 'Update WordPress NPM packages based on Gutenberg release') &&
        endsWith(github.event.workflow_run.head_commit.message, format('{0} {1}', 'github-actions', '<github-actions@github.com>')) &&
        github.event.workflow_run.head_commit.author.name == 'github-actions' &&
        github.event.workflow_run.head_commit.author.email == 'github-actions@github.com'
      run: gh pr merge "$PR_NUMBER" --auto --merge --delete-branch
      env:
        PR_NUMBER: ${{github.event.workflow_run.pull_requests[0].number}}
        GITHUB_TOKEN: ${{secrets.GITHUB_TOKEN}}
```

Reply:
> I'm not clear on how github.actor would be the same as the PAT user unless the specific GitHub user opened the PR.

With the above `github.actor` condition, it will always execute for one user that is trusted and will never commit with git `user.name` and `user.email` as github-action credentials.

Reply:
So is the intention that the workflow be triggered manually by a trusted and not automatically on a schedule?

Reply:
This workflow i.e. `auto-merge` is intended to triggered by `Test Measure & Build` workflow when all checks are successfully passed.  If the conditions are met for dependabot or Gutenberg update PRs then this workflow will run and merge the PRs.

Reply:
But the `Test, Measure, & Build` workflow runs for all pull requests, including those by contributors. So couldn't someone still craft a malicious PR that looks like it was authored by Dependabot or the Gutenberg update workflow that passes all the checks? Wouldn't this workflow then proceed to merge automatically?

Reply:
> Wouldn't this workflow then proceed to merge automatically?

This workflow will be triggered for all PR as it currently [does](https://github.com/ampproject/amp-wp/actions/workflows/auto-merge.yml). But it will process further only and only if it passes the condition on the jobs. This PR is having two jobs in `auto merge` workflow:

1. `dependabot` which will only runs if this condition passes -> `if: ${{ github.actor == 'dependabot[bot]' }}`
2. `gutenberg-packages` which will only runs if these two conditions passes:

```yml
1. if: ${{ github.actor == 'trusted-user' }}
2. if: |
    startsWith(github.event.workflow_run.head_commit.message, 'Update WordPress NPM packages based on Gutenberg release') &&
    endsWith(github.event.workflow_run.head_commit.message, format('{0} {1}', 'github-actions', '<github-actions@github.com>')) &&
    github.event.workflow_run.head_commit.author.name == 'github-actions' &&
    github.event.workflow_run.head_commit.author.email == 'github-actions@github.com'
```

Reply:
> So couldn't someone still craft a malicious PR that looks like it was authored by Dependabot or the Gutenberg update workflow that passes all the checks?

Someone can not get the dependabot or a `truested-user` actor context. 

Let's say a user `xyz` tried to create a malicious PR and committed it with git user as dependabot. When the workflows will run on his PR then the `github.actor` will be equal to `xyz`. So it's not possible for him to merge any malicious commit.